---
title: "Error Handling"
description: "Stop infinite retries with Terminal Errors."
icon: "triangle-exclamation"
---

Restate handles retries for failed invocations.

<Info>
    Check out the [Error Handling guide](/guides/error-handling) to learn more about how Restate handles transient errors, terminal errors, retries, and timeouts.
</Info>

## Retry strategies
By default, Restate does infinite retries with an exponential backoff strategy.

Check out the [error handling guide](/guides/error-handling) to learn how to customize this.

## Terminal Errors

For failures for which you do not want retries, but instead want the invocation to end and the error message
to be propagated back to the caller, you can throw a **terminal error**.

You can throw a `TerminalError` with an optional HTTP status code and a message anywhere in your handler, as follows:

```go {"CODE_LOAD::go/develop/errorhandling.go#here"} 
return restate.TerminalError(fmt.Errorf("Something went wrong."), 500)
```

You can catch terminal errors, and build your control flow around it.

<Info>
    When you throw a terminal error, you might need to undo the actions you did earlier in your handler to make sure that your system remains in a consistent state.
    Have a look at our [sagas guide](/guides/sagas) to learn more.
</Info>
